core/commands/block.go

```diff
@@ -185,3 +188,99 @@ func getBlockForKey(req cmds.Request, skey string) (blocks.Block, error) {
 	log.Debugf("ipfs block: got block with key: %q", b.Key())
 	return b, nil
 }
+
+var blockRmCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Remove IPFS block(s).",
+		ShortDescription: `
+'ipfs block put' is a plumbing command for removing raw ipfs blocks.
+It takes a list of base58 encoded multihashs to remove.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("hash", true, true, "Bash58 encoded multihash of block(s) to remove."),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("ignore-pins", "Ignore pins.").Default(false),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		ignorePins, _, err := req.Option("ignore-pins").Bool()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		hashes := req.Arguments()
+		keys := make([]key.Key, 0, len(hashes))
+		for _, hash := range hashes {
+			k := key.B58KeyDecode(hash)
+			keys = append(keys, k)
+		}
+		rdr, wtr := io.Pipe()
+		go func() {
+			pinning := n.Pinning
+			if ignorePins {
+				pinning = nil
+			}
+			err := rmBlocks(n.Blockstore, pinning, wtr, keys)
+			if err != nil {
+				wtr.CloseWithError(fmt.Errorf("Some blocks not deleted: %s", err))
+			} else {
+				wtr.Close()
+			}
+		}()
+		res.SetOutput(rdr)
```

Comment:
Looking this over again i see, Errors returned from readers in the commands lib are generally treated as more serious 'system level' errors (you will see an error message on the daemon). We don't normally use them for 'user' errors as mixing real errors and 'expected' errors makes handling them all tricky. If you disagree with this, we can rethink how that pipeline should work.

returning errors this way is handled by setting an http trailer on the request which makes handling them in client API libs more complicated.